### PR TITLE
improvements for TTL index removal queries

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Improvements for TTL index removal queries:

- do not have max query runtime affect TTL index removal queries
- do not favor certain databases/collections/shards in the removals, but give them more or less equal probability of being picked
- reduce the default amount of documents to be removed per collection (`--ttl.max-collection-removes` startup option`) from 1m to 40k. this allows different databases/collections/shards to be picked in a single TTL background thread iteration
- do not use an AQL query for removing documents from the TTL index. using an AQL query initiated on the DB server is not ideal because it will intermediate commit on followers but can be rolled back on leaders

- [x] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.11: -
  - [x] Backport for 3.10: this PR

#### Related Information

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/RestServer/TtlFeature.cpp [hunk 1]
```diff
@@ -302,6 +320,9 @@ class TtlThread final : public ServerThread<ArangodServer> {
             continue;
           }
 
+          // TODO: Filter out indexes that are currently still being built
+          // (checking isBuilding flag)
```
Comment (neunhoef) [inline arangod/RestServer/TtlFeature.cpp:324]:
```suggestion
          if (index->isHidden()) {
            // Filter out indexes that are currently still being built
            continue;
          }
```

File: arangod/RestServer/TtlFeature.cpp [hunk 2]
```diff
@@ -320,66 +341,160 @@ class TtlThread final : public ServerThread<ArangodServer> {
...
-                        << " documents for collection '" << collection->name()
-                        << "'";
-                    if (limitLeft >= removed) {
-                      limitLeft -= removed;
-                    } else {
-                      limitLeft = 0;
-                    }
-                  }
+
+            auto trx = std::make_unique<SingleCollectionTransaction>(
```
Comment (neunhoef) [inline arangod/RestServer/TtlFeature.cpp:422]: For a shard belonging to a smart edge collection, we must proceed as follows:
 1. If it is a `_local` shard: to what we do here normally (local removal with replication).
 2. If it is a `_from` shard: send removal as batch to the corresponding global edge collection on some coordinator.
 3. If it is a `_to` shard: skip it.

**Review by neunhoef — CHANGES_REQUESTED** [review]
This is great! I think it will solve our problems. But two fixes are still missing, see comments. We must skip indexes, which are still isBuilding, this is easy. Secondly, we must deal with smart edge collections correctly.

**Review by neunhoef — APPROVED** [review]
LGTM

File: arangod/RestServer/TtlFeature.cpp [hunk 3]
```diff
@@ -320,65 +356,252 @@ class TtlThread final : public ServerThread<ArangodServer> {
...
+              }
+
+              size_t count = 0;
+              for (auto it : VPackArrayIterator(val.slice())) {
+                VPackSlice error = it.get(StaticStrings::Error);
+                if (!error.isTrue()) {
+                  ++count;
+                  continue;
+                }
+              }
```
Comment (mpoeter) [inline arangod/RestServer/TtlFeature.cpp:517]:
```suggestion
              for (auto it : VPackArrayIterator(val.slice())) {
                VPackSlice error = it.get(StaticStrings::Error);
                if (!error.isTrue()) {
                  ++count;
                }
              }
```

**Review by mpoeter — APPROVED** [review]
I cannot even begin to express how much I dislike that this adds a lot more complexity (including duplicated code and mixing different abstraction levels) to a function that is already rather long. 

But I guess it works and has to be "quick" again.

File: arangod/RestServer/TtlFeature.cpp [hunk 4]
```diff
@@ -320,65 +356,252 @@ class TtlThread final : public ServerThread<ArangodServer> {
...
+                         "collection '"
+                      << collection->name() << "': " << res.errorMessage();
+                }
+                break;
+              }
+
+              OperationOptions opOptions;
+              opOptions.ignoreRevs = false;
+              opOptions.waitForSync = false;
+              // opOptions.silent = true;
```
Comment (mpoeter) [inline arangod/RestServer/TtlFeature.cpp:552]: Remove?
```suggestion
```

**Review by neunhoef — APPROVED** [review]
LGTM

**Pull request merged**